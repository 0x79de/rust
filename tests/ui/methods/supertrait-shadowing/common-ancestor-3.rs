//@ run-pass
//@ check-run-results

#![feature(supertrait_item_shadowing)]
#![warn(supertrait_item_shadowing_usage)]
#![allow(dead_code)]

trait A {
    fn hello(&self) {
        println!("A");
    }
}
impl<T> A for T {}

trait B {
    fn hello(&self) {
        println!("B");
    }
}
impl<T> B for T {}

trait C: A + B {
    fn hello(&self) {
        println!("C");
    }
}
impl<T> C for T {}

// `D` extends `C` which extends `B` and `A`

trait D: C {
    fn hello(&self) {
        println!("D");
    }
}
impl<T> D for T {}

fn main() {
    ().hello();
    //~^ WARN trait item `hello` from `D` shadows identically named item from supertrait
}
