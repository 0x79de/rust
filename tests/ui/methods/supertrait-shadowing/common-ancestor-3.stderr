warning: trait item `hello` from `D` shadows identically named item from supertrait
  --> $DIR/common-ancestor-3.rs:39:8
   |
LL |     ().hello();
   |        ^^^^^
   |
note: item from `D` shadows a supertrait item
  --> $DIR/common-ancestor-3.rs:32:5
   |
LL |     fn hello(&self) {
   |     ^^^^^^^^^^^^^^^
note: items from several supertraits are shadowed: `A`, `B`, and `C`
  --> $DIR/common-ancestor-3.rs:9:5
   |
LL |     fn hello(&self) {
   |     ^^^^^^^^^^^^^^^
...
LL |     fn hello(&self) {
   |     ^^^^^^^^^^^^^^^
...
LL |     fn hello(&self) {
   |     ^^^^^^^^^^^^^^^
note: the lint level is defined here
  --> $DIR/common-ancestor-3.rs:5:9
   |
LL | #![warn(supertrait_item_shadowing_usage)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: 1 warning emitted

